src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs

```diff
@@ -32,34 +32,93 @@ public Task BindModelAsync(ModelBindingContext bindingContext)
                 return TaskCache.CompletedTask;
             }
 
-            var mutableObjectBinderContext = new MutableObjectBinderContext()
-            {
-                ModelBindingContext = bindingContext,
-                PropertyMetadata = GetMetadataForProperties(bindingContext).ToArray(),
-            };
-
-            if (!(CanCreateModel(mutableObjectBinderContext)))
+            if (!(CanCreateModel(bindingContext)))
             {
                 return TaskCache.CompletedTask;
             }
 
-            return BindModelCoreAsync(bindingContext, mutableObjectBinderContext);
+            // Perf: separated to avoid allocating a state machine when we don't
+            // need to go async.
+            return BindModelCoreAsync(bindingContext);
         }
 
-        private async Task BindModelCoreAsync(
-            ModelBindingContext bindingContext,
-            MutableObjectBinderContext mutableObjectBinderContext)
+        private async Task BindModelCoreAsync(ModelBindingContext bindingContext)
         {
             // Create model first (if necessary) to avoid reporting errors about properties when activation fails.
-            var model = GetModel(bindingContext);
+            bindingContext.Model = GetModel(bindingContext);
+
+            foreach (var property in bindingContext.ModelMetadata.Properties)
+            {
+                if (!CanBindProperty(bindingContext, property))
+                {
+                    continue;
+                }
+
+                // ModelBindingContext.Model property values may be non-null when invoked via TryUpdateModel(). Pass
```

Comment:
Huh? `bindingContext.Model` will always be non-`null` given the `GetModel()` call just a few lines up.